doc/source/data/doc_code/custom_datasource.py

```diff
@@ -0,0 +1,127 @@
+# flake8: noqa
+
+# fmt: off
+# __read_single_query_start__
+from ray.data.block import Block
+
+# This connects to MongoDB, executes the query against it, converts the result into
+# Arrow format and returns the result as a Block.
+def _read_single_query(uri, database, collection, query, schema, kwargs) -> Block:
+    import pymongo
+    from pymongoarrow.api import aggregate_arrow_all
+
+    client = pymongo.MongoClient(uri)
+    return aggregate_arrow_all(
+        client[database][collection], query, schema=schema, **kwargs
+    )
+# __read_single_query_end__
+# fmt: on
+
+# fmt: off
+# __mongo_datasource_reader_start__
+from typing import Any, Dict, List, Optional
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+from ray.data.block import BlockMetadata
+
+class _MongoDatasourceReader(Reader):
+    # This is constructed by the MongoDatasource, which will supply these args
+    # about MongoDB.
+    def __init__(self, uri, database, collection, pipelines, schema, kwargs):
+        self._uri = uri
+        self._database = database
+        self._collection = collection
+        self._pipelines = pipelines
+        self._schema = schema
+        self._kwargs = kwargs
+
+    # Create a list of ``ReadTask``, one for each query (i.e. pipeline in MongoDB's
+    # specific context). Those tasks will be executed in parallel.
+    # The ``parallelism`` which is supposed to indicate how many ``ReadTask`` to
+    # return will have no effect here, since we map each query into a ``ReadTask``.
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        read_tasks: List[ReadTask] = []
+        for pipeline in self._pipelines:
+            metadata = BlockMetadata(
+                num_rows=None,
+                size_bytes=None,
+                schema=self._schema,
+                input_files=pipeline,
+                exec_stats=None,
+            )
+            read_task = ReadTask(
+                lambda uri=self._uri, database=self._database,
+                       collection=self._collection, pipeline=pipeline,
+                       schema=self._schema, kwargs=self._kwargs: [
```

Comment:
why do we construct these lambda args?

Reply:
like why not just do:

`ReadTask(lambda: _read_single_query(**your_args))`

Reply:
@richardliaw A few reasons:
1. Python lambdas use late-binding, so such closure captures would end up submitting `len(self._pipelines)` read tasks all with the same final `pipeline` query. We circumvent this by eagerly binding with default args.
2. Capturing `self` in the closure could add extra serialization bloat, e.g. it would send the full `self._pipelines` to every read task. In other datasources, this has led to serialization errors, object leaks, etc.